exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -558,20 +570,47 @@ public TestKit build() {
       checkCorrectServiceNumber(services.size());
       checkCorrectValidatorNumber();
       checkArtifactsDirectory();
-      TestKitServiceInstances[] testKitServiceInstances = mergeServiceSpecs();
+      TestKitServiceInstances testKitServiceInstances = prepareServicesConfiguration();
       return newInstance(testKitServiceInstances, nodeType, validatorCount,
           timeServiceSpec, artifactsDirectory);
     }
 
     /**
-     * Turn collection of service instances into a list of
-     * {@linkplain TestKitServiceInstances} objects for native to work with.
+     * Turn collections of artifacts and service instances into a
+     * {@linkplain TestKitServiceInstances} object for native to work with.
      */
-    private TestKitServiceInstances[] mergeServiceSpecs() {
+    private TestKitServiceInstances prepareServicesConfiguration() {
       checkDeployedArtifactsAreUsed();
-      return serviceArtifactFilenames.entrySet().stream()
-          .map(this::aggregateServiceSpecs)
-          .toArray(TestKitServiceInstances[]::new);
+      TestKitServiceInstances.Builder builder = TestKitServiceInstances.newBuilder();
+
+      for (Map.Entry<ServiceArtifactId, String> entry: serviceArtifactFilenames.entrySet()) {
+        ServiceArtifactId artifactId = entry.getKey();
+        Runtime.ArtifactSpec artifactSpec = Runtime.ArtifactSpec
```

Comment:
(Nit, here and elsewhere) We usually keep the `newBuilder` operation on the same line as class name,
unless there is a need to split lines:

```java
return Runtime.ArtifactId.newBuilder()
              .setRuntimeId(artifactId.getRuntimeId())
// ...
              .build();
```